src/transformation/context/visitors.ts

```diff
@@ -0,0 +1,157 @@
+    [ts.SyntaxKind.ImportDeclaration]: ts.ImportDeclaration;
+    [ts.SyntaxKind.ImportClause]: ts.ImportClause;
+    [ts.SyntaxKind.NamespaceImport]: ts.NamespaceImport;
+    [ts.SyntaxKind.NamedImports]: ts.NamedImports;
+    [ts.SyntaxKind.ImportSpecifier]: ts.ImportSpecifier;
+    [ts.SyntaxKind.ExportAssignment]: ts.ExportAssignment;
+    [ts.SyntaxKind.ExportDeclaration]: ts.ExportDeclaration;
+    [ts.SyntaxKind.NamedExports]: ts.NamedExports;
+    [ts.SyntaxKind.ExportSpecifier]: ts.ExportSpecifier;
+    [ts.SyntaxKind.MissingDeclaration]: ts.MissingDeclaration;
+    [ts.SyntaxKind.ExternalModuleReference]: ts.ExternalModuleReference;
+    [ts.SyntaxKind.JsxElement]: ts.JsxElement;
+    [ts.SyntaxKind.JsxSelfClosingElement]: ts.JsxSelfClosingElement;
+    [ts.SyntaxKind.JsxOpeningElement]: ts.JsxOpeningElement;
+    [ts.SyntaxKind.JsxClosingElement]: ts.JsxClosingElement;
+    [ts.SyntaxKind.JsxFragment]: ts.JsxFragment;
+    [ts.SyntaxKind.JsxOpeningFragment]: ts.JsxOpeningFragment;
+    [ts.SyntaxKind.JsxClosingFragment]: ts.JsxClosingFragment;
+    [ts.SyntaxKind.JsxAttribute]: ts.JsxAttribute;
+    [ts.SyntaxKind.JsxAttributes]: ts.JsxAttributes;
+    [ts.SyntaxKind.JsxSpreadAttribute]: ts.JsxSpreadAttribute;
+    [ts.SyntaxKind.JsxExpression]: ts.JsxExpression;
+    [ts.SyntaxKind.CaseClause]: ts.CaseClause;
+    [ts.SyntaxKind.DefaultClause]: ts.DefaultClause;
+    [ts.SyntaxKind.HeritageClause]: ts.HeritageClause;
+    [ts.SyntaxKind.CatchClause]: ts.CatchClause;
+    [ts.SyntaxKind.PropertyAssignment]: ts.PropertyAssignment;
+    [ts.SyntaxKind.ShorthandPropertyAssignment]: ts.ShorthandPropertyAssignment;
+    [ts.SyntaxKind.SpreadAssignment]: ts.SpreadAssignment;
+    [ts.SyntaxKind.EnumMember]: ts.EnumMember;
+    [ts.SyntaxKind.SourceFile]: ts.SourceFile;
+    // [ts.SyntaxKind.Bundle]: ts.Bundle;
+
+    // Not included above
+    [ts.SyntaxKind.TrueKeyword]: ts.BooleanLiteral;
+    [ts.SyntaxKind.FalseKeyword]: ts.BooleanLiteral;
+    [ts.SyntaxKind.NullKeyword]: ts.NullLiteral;
+    [ts.SyntaxKind.SuperKeyword]: ts.SuperExpression;
+    [ts.SyntaxKind.ThisKeyword]: ts.ThisExpression;
+};
+
+export type ExpressionLikeNode = ts.Expression | ts.QualifiedName | ts.ExternalModuleReference;
+export type StatementLikeNode = ts.Statement;
+export type VisitorResult<T extends ts.Node> = T extends ExpressionLikeNode
+    ? tstl.Expression
+    : T extends StatementLikeNode
+    ? OneToManyVisitorResult<tstl.Statement>
+    : T extends ts.SourceFile
+    ? tstl.Block
+    : OneToManyVisitorResult<tstl.Node>;
+
+export type FunctionVisitor<T extends ts.Node> = (node: T, context: TransformationContext) => VisitorResult<T>;
+export type ObjectVisitor<T extends ts.Node> = { transform: FunctionVisitor<T>; priority?: number };
+export type Visitor<T extends ts.Node> = FunctionVisitor<T> | ObjectVisitor<T>;
+export type Visitors = { [P in keyof NodesBySyntaxKind]?: Visitor<NodesBySyntaxKind[P]> };
+
+export interface TransformerPlugin {
```

Comment:
We could consider moving this to its own file if this interface grows, for now I don't know if there is a better location for this interface.